src/runmode-dpdk.c

```diff
@@ -0,0 +1,1237 @@
+}
+
+static int ConfigSetCopyIface(DPDKIfaceConfig *iconf, const char *entry_str)
+{
+    SCEnter();
+    int retval;
+
+    if (entry_str == NULL || entry_str[0] == '\0' || strcmp(entry_str, "none") == 0) {
+        iconf->out_iface = NULL;
+        SCReturnInt(0);
+    }
+
+    retval = rte_eth_dev_get_port_by_name(entry_str, &iconf->out_port_id);
+    if (retval < 0) {
+        SCLogError(SC_ERR_DPDK_CONF,
+                "Name of the copy interface (%s) for the interface %s is not valid, changing to %s",
+                entry_str, iconf->iface, DPDK_CONFIG_DEFAULT_COPY_INTERFACE);
+        iconf->out_iface = DPDK_CONFIG_DEFAULT_COPY_INTERFACE;
+    }
+
+    iconf->out_iface = entry_str;
+    SCReturnInt(0);
+}
+
+static int ConfigSetCopyMode(DPDKIfaceConfig *iconf, const char *entry_str)
+{
+    SCEnter();
+    if (entry_str == NULL) {
+        SCLogError(SC_ERR_INVALID_VALUE, "Interface %s has no copy mode specified, changing to %s ",
+                iconf->iface, DPDK_CONFIG_DEFAULT_COPY_MODE);
+        entry_str = DPDK_CONFIG_DEFAULT_COPY_MODE;
+    }
+
+    if (strcmp(entry_str, "none") != 0 && strcmp(entry_str, "tap") != 0 &&
+            strcmp(entry_str, "ips") != 0) {
+        SCLogError(SC_ERR_INVALID_VALUE,
+                "Copy mode \"%s\" is not one of the possible values (none|tap|ips) for interface "
+                "%s. Changing to %s",
+                entry_str, iconf->iface, DPDK_CONFIG_DEFAULT_COPY_MODE);
+        entry_str = DPDK_CONFIG_DEFAULT_COPY_MODE;
+    }
+
+    if (strcmp(entry_str, "none") == 0) {
+        iconf->copy_mode = DPDK_COPY_MODE_NONE;
+    } else if (strcmp(entry_str, "tap") == 0) {
+        iconf->copy_mode = DPDK_COPY_MODE_TAP;
+    } else if (strcmp(entry_str, "ips") == 0) {
+        iconf->copy_mode = DPDK_COPY_MODE_IPS;
+    }
+
+    SCReturnInt(0);
+}
+
+static int ConfigSetCopyIfaceSettings(DPDKIfaceConfig *iconf, const char *iface, const char *mode)
+{
+    SCEnter();
+    int retval;
+
+    retval = ConfigSetCopyIface(iconf, iface);
+    if (retval < 0)
+        SCReturnInt(retval);
+
+    retval = ConfigSetCopyMode(iconf, mode);
+    if (retval < 0)
+        SCReturnInt(retval);
+
+    if (iconf->copy_mode == DPDK_COPY_MODE_NONE) {
+        if (iconf->out_iface != NULL) {
+            SCFree((void *)iconf->out_iface);
+            iconf->out_iface = NULL;
+        }
+
+        SCReturnInt(0);
+    }
+
+    if (iconf->out_iface == NULL || strlen(iconf->out_iface) <= 0) {
+        SCLogError(SC_ERR_DPDK_CONF, "Copy mode enabled but interface not set");
+        SCReturnInt(-EINVAL);
+    }
+
+    if (iconf->copy_mode == DPDK_COPY_MODE_IPS)
+        SCLogInfo("DPDK IPS mode activated between %s and %s", iconf->iface, iconf->out_iface);
+    else if (iconf->copy_mode == DPDK_COPY_MODE_TAP)
+        SCLogInfo("DPDK IPS mode activated between %s and %s", iconf->iface, iconf->out_iface);
+
+    SCReturnInt(0);
+}
+
+static int ConfigLoad(DPDKIfaceConfig *iconf, const char *iface)
```

Comment:
At this function, checking if EAL parameter was specified and set to `primary` is too late. This can be done in InitEal() function in my opinion. DPDK apps start as primary processes by default. I can add the check directly there. Currently a nice thing that can be done when Suricata is running as a secondary process is to create a virtual port and send packets from the primary app there. It's something that Amaan has been doing in the previous PRs.